src/main/drivers/compass_hmc5883l.c

```diff
@@ -202,68 +200,64 @@ bool hmc5883lInit(void)
     delay(100);
     hmc5883lRead(magADC);
 
-    int validSamples = 0;
-    int failedSamples = 0;
-    while (validSamples < 10 && failedSamples < INIT_MAX_FAILURES) { // Collect 10 samples
+    int validSamples1 = 0;
+    int failedSamples1 = 0;
+    int saturatedSamples1 = 0;
+    while (validSamples1 < 10 && failedSamples1 < INITIALISATION_MAX_READ_FAILURES) { // Collect 10 samples
         i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_MODE, 1);
         delay(50);
         if (hmc5883lRead(magADC)) { // Get the raw values in case the scales have already been changed.
-            ++validSamples;
-            // Since the measurements are noisy, they should be averaged rather than taking the max.
-            xyz_total[X] += magADC[X];
-            xyz_total[Y] += magADC[Y];
-            xyz_total[Z] += magADC[Z];
             // Detect saturation.
             if (-4096 >= MIN(magADC[X], MIN(magADC[Y], magADC[Z]))) {
-                error_saturation = true;
-                bret = false;
-                break;              // Breaks out of the for loop.  No sense in continuing if we saturated.
+                ++saturatedSamples1;
+                ++failedSamples1;
+            } else {
+                ++validSamples1;
+                // Since the measurements are noisy, they should be averaged rather than taking the max.
+                xyz_total[X] += magADC[X];
+                xyz_total[Y] += magADC[Y];
+                xyz_total[Z] += magADC[Z];
+
             }
         } else {
-            ++failedSamples;
+            ++failedSamples1;
         }
         LED1_TOGGLE;
     }
 
-    if (failedSamples >= INIT_MAX_FAILURES) {
-        bret = false;
-    }
-
-    if (failedSamples > 0) {
-        error_read = true;
-    }
-
     // Apply the negative bias. (Same gain)
     i2cWrite(MAG_I2C_INSTANCE, MAG_ADDRESS, HMC58X3_R_CONFA, 0x010 + HMC_NEG_BIAS);   // Reg A DOR = 0x010 + MS1, MS0 set to negative bias.
-    validSamples = 0;
-    failedSamples = 0;
-    while (validSamples < 10 && failedSamples < INIT_MAX_FAILURES) { // Collect 10 samples
+    int validSamples2 = 0;
+    int failedSamples2 = 0;
+    int saturatedSamples2 = 0;
+    while (validSamples2 < 10 && failedSamples2 < INITIALISATION_MAX_READ_FAILURES) { // Collect 10 samples
```

Comment:
Yes, the above code certainly can be optimised. While tracking down a problem I make the minimum changes possible, leaving optimisation until later. Any simultaneous optimisation obscures things.